OCPBUGS-48469: Fix CoreDNS static pod bring-up on cloud platforms

sadasu opened the pull request:
Fixes: OCPBUGS-48469

**- What I did**
1. Updated CoreDNS Corefile errors for cloud platforms needing alternate in-cluster DNS when UserProvisionedDNS is enabled via the install-config
2. Updated the list of directories to include the location of the CoreDNS files for cloud platforms
3. Updated test data and unit tests for UserProvisionedDNS enabled on GCP.

**- How to verify it**
Set `UserProvisionedDNS` to `Enabled` for GCP via install-config and start installation

**- Description for the changelog**
Fixed issues with CoreDNS Corefile and template path for cloud platforms when UserProvisionedDNS is enabled.
Added test_data for GCP with all the UserProvisionedDNS configuration to better test this path.

sadasu commented:
/label acknowledge-critical-fixes-only

yuqi-zhang commented on pkg/controller/template/render.go:
```diff
@@ -222,6 +223,12 @@ func getPaths(config *RenderConfig, platformString string) []string {
+	// If this is a cloud platform with DNSType set to `ClusterHosted` with
+	// LB IPs provided, include path for their CoreDNS files
+	if cloudPlatformLoadBalancerIPState(*config) == availableLBIPState {
+		platformBasedPaths = append(platformBasedPaths, cloudPlatformAltDNS)
```
Before this, was anything processing the templates in `cloud-platform-alt-dns` ? Was is being done in bootstrap only?

sadasu replied:
It was being done on the Bootstrap node successfully and while debugging for the CoreDNS pod not starting on the master nodes, I found this issue.

sadasu commented:
/retest-required

mkowalski commented:
Ok from on-prem team. MCO can do the honours of merging if the order of merging templates makes sense

mkowalski commented on templates/common/cloud-platform-alt-dns/files/coredns.yaml:
```diff
@@ -40,6 +40,8 @@ contents:
         - "/config"
         - "--out-dir"
         - "/etc/coredns"
+        - "--resolvconf-path"
+        - "/var/run/NetworkManager/resolv.conf"
```
Why? Isn't passing /etc/resolv.conf enough?

sadasu replied:
I think /etc/resolv.conf is sufficient. I was going to ask you why /var/run/NetworkManager/resolv.conf was chosen for metal platform. Thanks for clarifying.

mkowalski reviewed (changes requested):
One new comment for the changed invocation of runtimecfg.

Plus a test will be good. MCO has all what's needed to test if the rendered config looks good.

sadasu commented:
> One new comment for the changed invocation of runtimecfg.
> 
> Plus a test will be good. MCO has all what's needed to test if the rendered config looks good.

Of course. Removed test temporarily because clusterbot was having merge conflicts while building an image with this and https://github.com/openshift/machine-config-operator/pull/4800. Now, I removed tests from the other one and moved it to WIP.

jianli-wei commented:
/label qe-approved

sadasu commented:
/retest-required

sadasu commented:
/test unit

sadasu commented:
/retest-required

sadasu commented:
/retest-required

sadasu commented:
/retest-required

yuqi-zhang reviewed (approved):
logically makes sense to me, thanks for adding a test

mkowalski commented on manifests/cloud-platform-alt-dns/coredns.yaml:
```diff
@@ -10,6 +10,7 @@ metadata:
     app: cloud-platform-infra-coredns
   annotations:
     target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
+    openshift.io/required-scc: privileged
```
@cybertron removed it long time ago for a reason. Why do we need it? It's just a DNS container, I can't see an obvious reason for it to be privileged, but I may be missing something subtle

sadasu replied:
@mko, @cybertron removed setting to run the CoreDNS pod's containers in privileged mode with https://github.com/openshift/machine-config-operator/pull/4443.

Change above is adding work done in https://github.com/openshift/machine-config-operator/pull/4504.

mkowalski commented on templates/common/cloud-platform-alt-dns/files/coredns.yaml:
```diff
@@ -13,6 +13,7 @@ contents:
         app: cloud-platform-infra-coredns
       annotations:
         target.workload.openshift.io/management: '{"effect": "PreferredDuringScheduling"}'
+        openshift.io/required-scc: privileged
```
Same as above, reverts old @cybertron's change

sadasu replied:
See https://github.com/openshift/machine-config-operator/pull/4830#discussion_r1991518715

mkowalski replied:
ACK even though it's now confusing. It means that our Pod requires privileged SCC but is not running as privileged itself. I hope we could understand it without PhD from auth, but not sure it's possible.

In any case, OK for me, sorry for the confusion

mkowalski commented on templates/common/cloud-platform-alt-dns/files/coredns.yaml:
```diff
@@ -24,6 +25,12 @@ contents:
       - name: conf-dir
         hostPath:
           path: "/etc/coredns"
+      - name: nm-resolv
+        hostPath:
+          path: "/var/run/NetworkManager"
```
Does it work as expected if Pod starts before this exists? Do we need to pass the whole directory or can we just pass the file? Name says `nm-resolv` so I would expect the path to point just to resolv.conf (unless this is a hack in case Pod starts before resolv.conf file exists, but after the directory has been created)

sadasu replied:
I don't think we have encountered the situation in our testing yet where the pod is unable to find `resolv.conf` at this location. It seemed like a good idea to not assume that the file is always going to be present when the pod is started.

mkowalski reviewed (changes requested):
Some comments, need @cybertron's review too

sadasu commented:
/retest-required

sadasu commented:
/hold cancel

sadasu commented:
/retest-required

ci/prow/e2e-gcp-op-single-node has passed earlier and no code changes were made after that.

sadasu commented:
/hold

sadasu commented:
Results of pre-merge testing with https://github.com/openshift/baremetal-runtimecfg/pull/345.
1. CoreDNS pod started successfully on masters with Corefile:
```
time="2025-03-20T03:57:17Z" level=info msg="Adding 10.0.0.6 as DNS Upstream"
time="2025-03-20T03:57:17Z" level=info msg="Adding 169.254.169.254 as DNS Upstream"
time="2025-03-20T03:57:17Z" level=info msg=". {"
time="2025-03-20T03:57:17Z" level=info msg="    errors"
time="2025-03-20T03:57:17Z" level=info msg="    bufsize 512"
time="2025-03-20T03:57:17Z" level=info msg="    health :18080"
time="2025-03-20T03:57:17Z" level=info msg="    forward . 10.0.0.6 169.254.169.254 {"
time="2025-03-20T03:57:17Z" level=info msg="        policy sequential"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="    cache 30"
time="2025-03-20T03:57:17Z" level=info msg="    reload"
time="2025-03-20T03:57:17Z" level=info msg="    template IN A jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:17Z" level=info msg="        match .*[.]apps.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:17Z" level=info msg="        answer \"{{ .Name }} 60 in {{ .Type }} 34.56.17.189\""
time="2025-03-20T03:57:17Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="    template IN AAAA jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:17Z" level=info msg="        match .*[.]apps.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:17Z" level=info msg="        "
time="2025-03-20T03:57:17Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="    template IN A jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:17Z" level=info msg="        match ^api.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:17Z" level=info msg="        answer \"{{ .Name }} 60 in {{ .Type }} 34.8.101.56\""
time="2025-03-20T03:57:17Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="    template IN AAAA jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:17Z" level=info msg="        match ^api.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:17Z" level=info msg="        "
time="2025-03-20T03:57:17Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="    template IN A jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:17Z" level=info msg="        match ^api-int.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:17Z" level=info msg="        answer \"{{ .Name }} 60 in {{ .Type }} 10.0.0.2\""
time="2025-03-20T03:57:17Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="    template IN AAAA jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:17Z" level=info msg="        match ^api-int.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:17Z" level=info msg="        "
time="2025-03-20T03:57:17Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="    hosts {"
time="2025-03-20T03:57:17Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:17Z" level=info msg="    }"
time="2025-03-20T03:57:17Z" level=info msg="}"
time="2025-03-20T03:57:17Z" level=info
time="2025-03-20T03:57:17Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
```

sadasu commented:
2. CoreDNS pod continuing to run with updated Corefile:
```
time="2025-03-20T03:57:18Z" level=info msg="Adding 10.0.0.6 as DNS Upstream"
time="2025-03-20T03:57:18Z" level=info msg="Adding 169.254.169.254 as DNS Upstream"
time="2025-03-20T03:57:18Z" level=error msg="Failed to get node list: Get \"https://api-int.jiwei-0320b.qe.gcp.devcluster.openshift.com:6443/api/v1/nodes\": dial tcp: lookup api-int.jiwei-0320b.qe.gcp.devcluster.openshift.com on 169.254.169.254:53: no such host"
time="2025-03-20T03:57:48Z" level=info msg="Adding 10.0.0.6 as DNS Upstream"
time="2025-03-20T03:57:48Z" level=info msg="Adding 169.254.169.254 as DNS Upstream"
time="2025-03-20T03:57:49Z" level=info msg="Node change detected, rendering Corefile" Node Addresses="[{10.0.0.5 jiwei-0320b-rlfw2-master-0 false} {10.0.0.3 jiwei-0320b-rlfw2-master-1 false} {10.0.0.6 jiwei-0320b-rlfw2-master-2 false}]"
time="2025-03-20T03:57:49Z" level=info msg=". {"
time="2025-03-20T03:57:49Z" level=info msg="    errors"
time="2025-03-20T03:57:49Z" level=info msg="    bufsize 512"
time="2025-03-20T03:57:49Z" level=info msg="    health :18080"
time="2025-03-20T03:57:49Z" level=info msg="    forward . 10.0.0.6 169.254.169.254 {"
time="2025-03-20T03:57:49Z" level=info msg="        policy sequential"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="    cache 30"
time="2025-03-20T03:57:49Z" level=info msg="    reload"
time="2025-03-20T03:57:49Z" level=info msg="    template IN A jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:49Z" level=info msg="        match .*[.]apps.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        answer \"{{ .Name }} 60 in {{ .Type }} 34.56.17.189\""
time="2025-03-20T03:57:49Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="    template IN AAAA jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:49Z" level=info msg="        match .*[.]apps.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        "
time="2025-03-20T03:57:49Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="    template IN A jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:49Z" level=info msg="        match ^api.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        answer \"{{ .Name }} 60 in {{ .Type }} 34.8.101.56\""
time="2025-03-20T03:57:49Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="    template IN AAAA jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:49Z" level=info msg="        match ^api.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        "
time="2025-03-20T03:57:49Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="    template IN A jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:49Z" level=info msg="        match ^api-int.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        answer \"{{ .Name }} 60 in {{ .Type }} 10.0.0.2\""
time="2025-03-20T03:57:49Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="    template IN AAAA jiwei-0320b.qe.gcp.devcluster.openshift.com {"
time="2025-03-20T03:57:49Z" level=info msg="        match ^api-int.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        "
time="2025-03-20T03:57:49Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="    hosts {"
time="2025-03-20T03:57:49Z" level=info msg="        10.0.0.5 jiwei-0320b-rlfw2-master-0 jiwei-0320b-rlfw2-master-0.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        10.0.0.3 jiwei-0320b-rlfw2-master-1 jiwei-0320b-rlfw2-master-1.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        10.0.0.6 jiwei-0320b-rlfw2-master-2 jiwei-0320b-rlfw2-master-2.jiwei-0320b.qe.gcp.devcluster.openshift.com"
time="2025-03-20T03:57:49Z" level=info msg="        fallthrough"
time="2025-03-20T03:57:49Z" level=info msg="    }"
time="2025-03-20T03:57:49Z" level=info msg="}"
time="2025-03-20T03:57:49Z" level=info
time="2025-03-20T03:57:49Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
``
```

sadasu commented:
/retest-required

issues were related to pulling images from quay

sadasu commented:
/retest-required

sadasu commented:
/retest-required

sadasu commented:
ci/prow/bootstrap-unit - has been prema-failing.

mkowalski commented:
/lgtm

sadasu commented:
/retest-required

sadasu commented:
/retest-required

sadasu commented:
/retest-required

patrickdillon commented:
/skip
/retest-required

yuqi-zhang commented:
The 3 remaining tests are blocked on general CI issues. If this can wait until after feature complete (tomorrow), let's wait for CI to stablize. otherwise, please let me know and I can override

patrickdillon commented:
/lgtm

bumping after rebase

sadasu commented:
> The 3 remaining tests are blocked on general CI issues. If this can wait until after feature complete (tomorrow), let's wait for CI to stablize. otherwise, please let me know and I can override

@yuqi-zhang I would like this to merge before FF. This feature is on track to be GA in this release minus these hiccups. Thanks for the offer to override. I might take you up on it.

sadasu commented:
Is the test /test 4.12-upgrade-from-stable-4.11-images relevant to 4.19 and this code ? If not could we please override it?

sadasu commented:
/override ci/prow/okd-images

Perma-failing job.

sadasu commented:
/retest-required

sadasu commented:
> The 3 remaining tests are blocked on general CI issues. If this can wait until after feature complete (tomorrow), let's wait for CI to stablize. otherwise, please let me know and I can override

@yuqi-zhang CI jobs are slowly crawling back but not sure how long it would take before they stabilize. Could you override the 2 aws jobs that are still having issues? I do want this to merged before feature complete.

sadasu commented:
The AWS jobs are failing due to us hitting quota limits.

```
level=error msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed during pre-provisioning: failed to create IAM roles: failed to create IAM master role: failed to create master role: LimitExceeded: Cannot exceed quota for RolesPerAccount: 1000
level=error msg=	status code: 409, request id: 5c4a4e45-7f53-4aab-b20e-1ec748905186
Installer exit with code 4
```

patrickdillon commented:
/skip
/retest-required

gpei commented:
/retest-required

sadasu commented:
/retest-required

zaneb commented:
/retest-required
